product_variant_template_reassign/wizards/reassign_variant.py

```diff
@@ -0,0 +1,133 @@
+                    ("id", "!=", wiz.origin_product_template_id.id),
+                    ("attribute_line_ids", "!=", False),
+                ]
+            )
+
+    @api.depends("target_product_template_id")
+    def _compute_allowed_attribute_value_ids(self):
+        self.allowed_attribute_value_ids = False
+        for wiz in self.filtered("target_product_template_id"):
+            wiz.allowed_attribute_value_ids = (
+                self.env["product.attribute.value"].search(
+                    [
+                        (
+                            "attribute_id",
+                            "in",
+                            wiz.target_product_template_id.attribute_line_ids.attribute_id.ids,
+                        ),
+                        ("attribute_id.create_variant", "=", "always"),
+                    ]
+                )
+                - wiz.target_product_template_id.attribute_line_ids.value_ids
+            )
+
+    @api.depends("target_product_template_id")
+    def _compute_attribute_value_ids(self):
+        """Whenever the target changes we must wipe the attribute values"""
+        self.attribute_value_ids = False
+
+    def reassign(self):
+        """No way back! The potential new variant from the selected attributes is
+        merged with the origin variant"""
+        # 1. Create variant
+        existing_variants = self.target_product_template_id.product_variant_ids
+        for attribute_value in self.attribute_value_ids:
+            self.target_product_template_id.attribute_line_ids.filtered(
+                lambda x: x.attribute_id == attribute_value.attribute_id
+            ).value_ids += attribute_value
+        new_variant = (
+            self.target_product_template_id.product_variant_ids - existing_variants
+        )
+        if not new_variant:
+            raise UserError(
+                _(
+                    "The attributes selected didn't generate a variant in the target template"
+                )
+            )
+        try:
+            new_variant.ensure_one()
+        except ValueError:
+            raise UserError(
+                _(
+                    "The attributes selected generate more than one variant. "
+                    "Refine your configuration"
+                )
+            ) from None
+        # 2. Merge existing variant into the new one. Do it with SQL to avoid ORM locks
+        merge_records(
+            self.env,
+            "product.product",
+            self.origin_product_template_id.product_variant_id.ids,
+            new_variant.id,
+            method="sql",
+        )
+        # 3. Merge origin template into the target template. Do it with SQL to avoid ORM
+        # locks.
+        merge_records(
+            self.env,
+            "product.template",
+            self.origin_product_template_id.ids,
+            self.target_product_template_id.id,
+            method="sql",
+        )
+        return {
```

Comment:
Please drop a comment in the thread saying that it has been merged for tracking purposes.

Reply:
done